lib/aiken-design-patterns/linked-list/ordered.ak

```diff
@@ -0,0 +1,251 @@
+use cardano/assets.{AssetName, flatten}
+use cardano/transaction.{InlineDatum, Input, Output}
+
+pub fn list_state_transition(node_mint: Dict<AssetName, Int>) {
+  unordered.list_state_transition(node_mint)
+}
+
+pub fn modify_data(
+  node_mint: Dict<AssetName, Int>,
+  own_input: Input,
+  own_output: Output,
+  node_nft_policy_id: ByteArray,
+  node_nft_asset_name: ByteArray,
+) -> Bool {
+  unordered.modify_data(
+    node_mint,
+    own_input,
+    own_output,
+    node_nft_policy_id,
+    node_nft_asset_name,
+  )
+}
+
+pub fn is_root_node(node_key: NodeKey) {
+  unordered.is_root_node(node_key)
+}
+
+pub fn is_last_node(node_link: NodeKey) {
+  unordered.is_last_node(node_link)
+}
+
+pub fn is_empty_list(node_key: NodeKey, node_link: NodeKey) {
+  unordered.is_empty_list(node_key, node_link)
+}
+
+pub fn key_added(key: NodeKey, node_mint: Dict<AssetName, Int>) {
+  unordered.key_added(key, node_mint)
+}
+
+pub fn key_removed(key: NodeKey, node_mint: Dict<AssetName, Int>) {
+  unordered.key_removed(key, node_mint)
+}
+
+pub fn is_member(
+  k: NodeKey,
+  node_key: NodeKey,
+  node_address: Address,
+  linked_list_address: Address,
+) {
+  unordered.is_member(k, node_key, node_address, linked_list_address)
+}
+
+pub fn is_not_member(
+  k: NodeKey,
+  node_key: NodeKey,
+  node_link: NodeKey,
+  node_address: Address,
+  linked_list_address: Address,
+) {
+  or {
+    node_address != linked_list_address,
+    and {
+      k != Empty,
+      {
+        expect Key(actual_key) = k
+        and {
+          or {
+            unordered.is_root_node(node_key),
+            {
+              expect Key(actual_node_key) = node_key
+              bytearray.compare(actual_node_key, actual_key) == Less
+            },
+          },
+          or {
+            unordered.is_last_node(node_link),
+            {
+              expect Key(actual_node_link) = node_link
+              bytearray.compare(actual_key, actual_node_link) == Less
+            },
+          },
+        }
+      },
+    },
+  }
+}
+
+//Application code must ensure that this action can happen only once
+pub fn init(
+  node_outputs: List<Output>,
+  node_mint: Dict<AssetName, Int>,
+  node_cs: ByteArray,
+) {
+  unordered.init(node_outputs, node_mint, node_cs)
+}
+
+pub fn deinit(
+  node_in_outputs: List<Output>,
+  node_mint: Dict<AssetName, Int>,
+  node_cs: ByteArray,
+) {
+  unordered.deinit(node_in_outputs, node_mint, node_cs)
+}
+
+pub fn prepend(
+  key_to_prepend: NodeKey,
+  prepended_node_output_index: Int,
+  anchor_node_output_index: Int,
+  node_in_outputs: List<Output>,
+  node_outputs: List<Output>,
+  node_mint: Dict<AssetName, Int>,
+) {
+  //We know it's only 1 from unordered_prepend()
+  expect Some(anchor_node_input) = list.head(node_in_outputs)
+  let anchor_node_input_raw_key_and_link =
+    get_raw_key_and_link(anchor_node_input)
+  expect Some(prepended_node_output) =
+    list.at(node_outputs, prepended_node_output_index)
+  and {
```

Comment:
Semantically, the handlers should crash the entire validator if their conditions don't hold.
They shouldn't merely return a False value.

Also, I find it much harder to compare the implementation code to the spec when the boolean conditions are moved to a monster `and` block at the end. It's better to replace them with non-assignment `expect` statements at similar positions as in the spec.

Reply:
Resolved by https://github.com/Anastasia-Labs/aiken-design-patterns/pull/12